### PR TITLE
[Bugfix] Make DP padding optional in coordinate_batch_across_dp

### PR DESCRIPTION
**Pull request by SageMoore**
## Purpose

`coordinate_batch_across_dp` currently assumes all DP ranks require padding to the maximum token count across ranks. This results in excessive padding during eager-mode prefill runs, and assertion failures in the `EagleProposer`.

Currently, main trips the following assertion when running `eagle618/deepseek-v3-random`:
https://github.com/vllm-project/vllm/blob/main/vllm/forward_context.py#L270

The previous implementation of `get_dp_padding` handled this by disabling DP padding when eager mode was enabled. 

This PR replicates that functionality by adding an `allow_dp_padding` parameter to `coordinate_batch_across_dp`:

- When `allow_dp_padding=False`: `num_tokens_after_padding` preserves each rank's original token count (no padding applied)
- When `allow_dp_padding=True`: `num_tokens_after_padding` is set to the maximum token count across all ranks (existing behavior)

Note: when `coordinate_batch_across_dp` decides to use microbatching, DP padding will be applied unconditionally regardless of the `allow_dp_padding` setting.

## Test Plan
lm_eval

## Test Result
`VLLM_ALL2ALL_BACKEND=deepep_low_latency vllm serve --model="deepseek-ai/DeepSeek-V2-Lite" --trust-remote-code --data-parallel-size 2 --enable-expert-parallel --enforce-eager`
 
```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.3567|±  |0.0277|
|     |       |strict-match    |     5|exact_match|↑  |0.3533|±  |0.0276|
```

`VLLM_ALL2ALL_BACKEND=deepep_low_latency vllm serve --model="deepseek-ai/DeepSeek-V2-Lite" --trust-remote-code --data-parallel-size 2 --enable-expert-parallel --enable-dbo --dbo-decode-token-threshold 4 --enforce-eager`

```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.3633|±  |0.0278|
|     |       |strict-match    |     5|exact_match|↑  |0.3600|±  |0.0278|
````

### CODE REVIEW

File: vllm/v1/worker/gpu_model_runner.py [hunk 1]
```diff
@@ -1170,13 +1170,18 @@ def _prepare_inputs(
         uniform_decode = (
             max_num_scheduled_tokens == self.uniform_decode_query_len
         ) and (total_num_scheduled_tokens == num_reqs * max_num_scheduled_tokens)
+
+        # Disable DP padding when running eager
+        allow_dp_padding = self.compilation_config.cudagraph_mode != CUDAGraphMode.NONE
```
Comment (LucasWilkinson) [inline vllm/v1/worker/gpu_model_runner.py:1187]: wouldn't we want DP padding for MoE chunking even in eager? do you know why this done? if not can you add a comment explaining you are preserving existing behavior?

File: vllm/v1/worker/dp_utils.py [hunk 2]
```diff
@@ -72,90 +74,141 @@ def _post_process_ubatch(tensor: torch.Tensor) -> bool:
...
-    )
+    should_dp_pad = bool(torch.all(tensor[3] == 1).item())
+
+    # DP ranks should all have the same value for should_attempt_dp_padding.
+    assert should_attempt_dp_padding == should_dp_pad
 
+    # Check conditions for microbatching
     should_ubatch = _post_process_ubatch(tensor)
 
+    if should_ubatch and not should_dp_pad:
```
Comment (LucasWilkinson) [inline vllm/v1/worker/dp_utils.py:135]: I think micro-batching might only require padding when using CGs or a a2a that needs chunking; but we can investigate this in a follow up :+1:

File: vllm/v1/worker/dp_utils.py [hunk 3]
```diff
@@ -72,90 +74,141 @@ def _post_process_ubatch(tensor: torch.Tensor) -> bool:
...
+        # If DP padding is enabled, ensure that each rank is processing the same number
+        # of tokens
+        max_num_tokens = int(num_tokens_across_dp.max().item())
+        return torch.tensor(
+            [max_num_tokens] * len(num_tokens_across_dp),
+            device="cpu",
+            dtype=torch.int32,
+        )
+    else:
+        return num_tokens_across_dp
```
[the remark this answers is not in the document]
Comment (LucasWilkinson) [reply vllm/v1/worker/dp_utils.py]: @SageMoore did you test eager + DP > 1? is this true?

**Review by LucasWilkinson — COMMENTED** [review]
Thanks for fixing this! Left a couple comments

File: vllm/v1/worker/gpu_model_runner.py (hunk 1 above)
Comment (SageMoore) [reply vllm/v1/worker/gpu_model_runner.py:1187]: When we do the chunking we look at the max number of tokens across all ranks.

https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/layers/fused_moe/layer.py#L2166

I'll add the comment from the original code.

File: vllm/v1/worker/dp_utils.py (hunk 2 above)
Comment (SageMoore) [reply vllm/v1/worker/dp_utils.py:135]: Definitely. Chunking is the big blocker here and we may be able to do something similar to the `max_tokens_across_dp_cpu` to get around this when running with DBO.

File: vllm/v1/worker/dp_utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (SageMoore) [reply vllm/v1/worker/dp_utils.py]: It is not true, but still good to copy this tensor over to CPU :).

**Review by tlrmchlsmth — APPROVED** [review]
